Blog post: Shared State in Functional Programming

gvolpe opened the pull request:
First shared in this [gist](https://gist.github.com/gvolpe/d4facceedf12ab7870934cbf9d6174e9) but here I added a few other things after getting some feedback from Reddit :)

Also thanks to @tpolecat for pushing me to write it as a `Typelevel` blog post.

larsrh commented:
@adelbertc Mind giving this  a review?

adelbertc commented:
@gvolpe @larsrh I will be sure to do it tomorrow!

gvolpe commented:
Great, thanks @adelbertc !

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
```diff
@@ -0,0 +1,372 @@
+
+---
+
+Newcomers to Functional Programming are often very confused about the proper way to share state without breaking purity and end up having a mix of pure and impure code that defeats the purpose of having pure FP code in the first place.
```
No need to capitalize "functional programming" and consider parenthesizing (FP) next to it so your future use of FP is clear

adelbertc replied:
This might be a good article to cite https://queue.acm.org/detail.cfm?id=2611829

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
```diff
@@ -0,0 +1,372 @@
+
+##### What is Referential Transparency?
+
+Being able to substitute an expression by its bound value without changing the meaning of the program.
```
The use of "bound value" might suggest this only applies to variable assignment/capture, in general RT says a given expression *e* is referentially transparent if you can evaluate it and substitute its value in **any** larger expression it appears in without changing the meaning of the larger expression.

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
````diff
@@ -0,0 +1,372 @@
+(123, 123)
+```
+
+No! The program remains the same. We call this property *Referential Transparency*. How about this?
````
No need to capitalize referential transparency.

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
````diff
@@ -0,0 +1,372 @@
+(println("hey"), println("hey"))
+```
+
+M'kay... The meaning of the program has now changed. The first program prints "hey" once whereas the second program prints "hey" twice. This is not `Referential Transparent`.
````
No need to capitalize

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
```diff
@@ -0,0 +1,372 @@
+import cats.effect._
+import cats.effect.concurrent.Ref
+import cats.instances.list._
+import cats.syntax.all._
```
You can replace these last 2 imports with just `import cats.implicits._`

gvolpe replied:
Yes, but tbh I prefer to be explicit about the imports

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
````diff
@@ -0,0 +1,372 @@
+
+So this is how we might decide to start writing our code:
+
+```scala
````
Consider using `tut` here so it actually goes through the type checker in CI http://tpolecat.github.io/tut/

Going to stop writing this for the rest of the review, but there's a couple code blocks below that can also benefit

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
````diff
@@ -0,0 +1,372 @@
+
+No! The program remains the same. We call this property *Referential Transparency*. How about this?
+
+```scala
````
Consider using `tut`

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
````diff
@@ -0,0 +1,372 @@
+
+An example is worth thousand words. Consider this:
+
+```scala
````
Consider using `tut`

larsrh replied:
There's even a `tut` header in the file :thinking:

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
````diff
@@ -0,0 +1,372 @@
+
+If we now substitute the expression by its bound value, does the meaning of the program change?
+
+```
````
Consider using `tut`

gvolpe replied:
I'm :-1: on using `tut` for a blog post. I think it's great for docs to keep it updated but not really for a blog post. 

The reason being that to preview this blog post I must compile all the existent blog posts and we might have divergence in the future when updating dependencies and an old post that was completely valid back in time is no longer valid, even if we force it to compile, because also the sentences will need to be updated.

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
````diff
@@ -0,0 +1,372 @@
+
+Let's perform susbtitution once again:
+
+```scala
````
Consider using `tut`

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
```diff
@@ -0,0 +1,372 @@
+
+As I mentioned in one of the sections above, the creation of `Ref[F, A]` is side-effectful. So what does this mean? Does it write to disk? Does it perform HTTP Requests? Not exactly.
+
+It all comes down to wanting to keep the property of *Referential Transparency* while sharing and mutating state. So let's again put up an example to follow up along with some explanation:
```
no need to capitalize

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
```diff
@@ -0,0 +1,372 @@
+
+## Applying the technique in other libraries
+
+Although in the example above we only see how it's done with the `Cats Effect` library, this principle expands to other FP libraries as well.
```
`cats-effect`

adelbertc commented on posts/2018-06-07-shared-state-in-fp.md:
```diff
@@ -0,0 +1,372 @@
+
+> In the specific case of state sharing, this gives rise to a really nice property: since the only way to share is passing things as an argument, *the regions of sharing are exactly the same of your call graph*, so you transform an important aspect of the behaviour ("who shares this state?") into a straightforward syntactical property ("what methods take this argument"?). This makes shared state  ...
+
+In simple terms, remind yourself about this: **"FlatMap once and pass the reference as an argument!"**
```
`flatMap`

gvolpe commented:
Thanks for the feedback @adelbertc ! Maybe I should leave the "Referential Transparency" section out and aim for the second type of audience as the [original post](https://gist.github.com/gvolpe/d4facceedf12ab7870934cbf9d6174e9). I'll be addressing the comments soon.

gvolpe commented:
Just give a bit more of context, this post is aimed for those that are currently making use of `IO` or `Task`, understand basic operations such as `flatMap` and `map` or even more but still don't feel super comfortable doing pure FP and end up mixing it with mutable state and a few `unsafeRunSync()`.

SethTisue commented:
> Just to give a bit more of context, this post is aimed for those that are currently making use of IO or Task, understand basic operations such as flatMap and map or even more but still don't feel super comfortable doing pure FP and end up mixing it with mutable state and a few unsafeRunSync().

succinctly put. say so at the top of the blog post itself...?

gvolpe commented:
@SethTisue I think that would probably be too much but I did add links to `IO` and `Ref`, and subtly suggested that some knowledge of Cats Effect is required in order to understand the examples in the Getting Started section.

larsrh commented:
@gvolpe 

> The reason being that to preview this blog post I must compile all the existent blog posts

You'll have to do that anyway, even if this post doesn't use it.

> we might have divergence in the future when updating dependencies and an old post that was completely valid back in time is no longer valid, even if we force it to compile, because also the sentences will need to be updated.

No, that's not a risk. The `tut` header specifies a fixed set of dependencies per file.

In any case, there's no need for your post to use it, but in that case, please remove the header.

gvolpe commented:
Thanks for correcting me @larsrh , my bad, didn't pay attention to the header. I'll try to get it compiling with `tut` as suggested :+1:

gvolpe commented:
I actually caught some bad code using `tut`, good advice :smile: In some parts I didn't use it either because of some limitations of `tut` (eg `private class`) or for convenience (eg. the `http4s` example at the end).

larsrh commented:
@adelbertc Did @gvolpe address all your comments? If so, please update your review to green.

larsrh commented:
@gvolpe Thanks for your contribution!
